packages/compiler/core/program.ts

```diff
@@ -59,133 +69,23 @@ export interface Program {
   reportDiagnostic(diagnostic: Diagnostic): void;
   reportDiagnostics(diagnostics: readonly Diagnostic[]): void;
   reportDuplicateSymbols(symbols: SymbolTable | undefined): void;
-  enableProjections(projections: ProjectionApplication[], startNode?: Type): Projector;
-  disableProjections(): void;
-  currentProjector?: Projector;
+  getGlobalNamespaceType(): Namespace;
 }
 
 interface EmitterRef {
   emitter: Emitter;
   options: EmitterOptions;
 }
 
-class StateMap<V> implements Map<Type, V> {
-  private internalState = new Map<undefined | Projector, Map<Type, V>>();
-  constructor(public program: Program, public key: symbol) {}
-
-  has(t: Type) {
-    return this.dispatch(t)?.has(t) ?? false;
-  }
-
-  set(t: Type, v: any) {
-    this.dispatch(t).set(t, v);
-    return this;
-  }
-
-  get(t: Type) {
-    return this.dispatch(t).get(t);
-  }
-
-  delete(t: Type) {
-    return this.dispatch(t).delete(t);
-  }
-
-  forEach(cb: (value: V, key: Type, map: Map<Type, V>) => void, thisArg?: any) {
-    this.dispatch().forEach(cb, thisArg);
-    return this;
-  }
-
-  get size() {
-    return this.dispatch().size;
-  }
-
-  clear() {
-    return this.dispatch().clear();
-  }
-
-  entries() {
-    return this.dispatch().entries();
-  }
-
-  values() {
-    return this.dispatch().values();
-  }
-
-  keys() {
-    return this.dispatch().keys();
-  }
-
-  [Symbol.iterator]() {
-    return this.entries();
-  }
-
-  [Symbol.toStringTag] = "StateMap";
-
-  dispatch(keyType?: Type): Map<Type, V> {
-    const key = keyType ? keyType.projector : this.program.currentProjector;
```

Comment:
Yeah that was the one part that I was talking about. I don't actually need to remove that, that was my initial thought that this was what was the core piece of the currentProjector but I am pretty sure I could revert that. 

However I am wondering if we have that, it would hide some potential issue(that we already have)

```ts
// Either of those works
getData(program, myProjectedEntity) === getData(projectedProgram, myProjectedEntity)

// But now if I have something that is not tied to an entity I have to make sure that I use the right prorgram
getServiceDetails(program) !== getServiceDetails(projectedProgram) 
```
Though I am not sure this is much better if you have `getData(program, myProjectedEntity)` return no value because the data is in the wrong place

Reply:
I wonder if we can tie getServiceDetails to an entity. Maybe the global namespace? Do we have other state that isn't tied to an entity?

Reply:
`getAllRoutes` as well, I think this is a common pattern when you want to keep index

Reply:
Just note here too that I like the realm idea so despite not belonging to this PR completely I could provide some of the setup here like adding a `realmId` to the program and updating this state to use the realmId as key.

Reply:
Sounds good. @bterlson suggested we all sync up so I put some time on the calendar.